agenthub/codeact_agent/prompt.py

```diff
@@ -42,6 +42,8 @@
 import IPython
 IPython.Application.instance().kernel.do_shutdown(True)  # Restart the kernel
 </execute_ipython>
+When the assistant has completed the user's latest request, the assistant MUST exit the environment by responding:
+<finish></finish>
```

Comment:
I generally like the idea of having a way for the agent to signal the completion of the "current sub-task" by emitting [FINISH], but if there's no command in between brackets, maybe we can simplify it a bit by changing it to `<finish>` tag?

Generally, I'm hesitant to add any more tags to the prompt -> more tags mean more instability (e.g., more things the model needs to pay attention to - means harder to do this with smaller LLMs). But i get this can be good for improving the UX (e.g., user know exactly what the model is doing now).

I'm actually wondering if it is possible to do it "outside" of CodeAct's prompt. e.g., we chatted about a potential state indicator/parser earlier, that can be the same or a separate (cheaper LLM) interprets an agent's action into a "message" (e.g., OpenDevin is reading readme, OpenDevin is waiting for user input) -> It is essentially a multi-class classification problem, INPUT a trajectory (all messages until this point + a question "is this task finished/XXX/YYY?", OUTPUT a current state (the current task is finished, OR the current task is awaiting user input).

IMO that can be a potentially more general way of getting these states (for all the agents) without directly modifying agent's prompts, which may or may not hurt the performance of agent, WDYT?

Reply:
Sounds like a promising line of research! Some possible issues I can foresee:
* You'd need _two_ prompts per step, which could get slow/expensive
  * If we wanted to use a cheaper LLM for the interpretation step, we'd need the infra to set up and configure that
* We'd need to eval the "meta agent" of CodeAct + Interpreter to understand how good OpenDevin is

Another thing I tried was using the existing `exit` command, but nothing I tried got the agent to use it. Maybe you have ideas there? 

It'd be worth evaluating what the performance impact of new tags are, so we have a quantitative sense here.

Reply:
Yeah i totally agree! though i guess the role of the "interpreter" would be smth more like a "ui-friendly" thing, which its inaccuracy would not really affect the performance of the agent.

That said, we definitely need to evaluate the performance if we want to add extra tags -- can we defer this to after SWE-Bench is merged? I'm pretty swamped this week due to the conference, but will try to get that merged around the beginning of next week.

Reply:
> smth more like a "ui-friendly" thing, which its inaccuracy would not really affect the performance of the agent

I actually disagree here. Ultimately what we're chasing is a great software agent, not SWE-bench scores--the latter is only a rough measure of the former. And "knowing when you're done" is definitely an important skill!

If the "are we done" agent is bad (e.g. it says "done" before the task is done), then our evals need to show that, because it means OpenDevin is getting worse, even if a sub-component (CodeAct) is OK.

Reply:
And yes--let's definitely hold off on changing CodeAct until we can eval!